web/src/classic/core/mantle/data/shapefile/parseZip.ts

```diff
@@ -0,0 +1,89 @@
+import JSZip from "jszip";
+import proj4 from "proj4";
+
+import { parseDbf } from "./parseDbf";
+import { parseShp } from "./parseShp";
+import { combine } from "./shapefile";
+
+function toBuffer(b: ArrayBuffer | Buffer): Buffer {
+  if (!b) {
+    throw new Error("forgot to pass buffer");
+  }
+  if (Buffer.isBuffer(b)) {
+    return b;
+  }
+  if (b instanceof ArrayBuffer) {
+    return Buffer.from(b);
+  }
+  throw new Error("Unknown buffer type");
+}
+
+export async function parseZip(
+  buffer: ArrayBuffer | Buffer,
+  whiteList?: string[],
+): Promise<GeoJSON.GeoJSON | GeoJSON.GeoJSON[]> {
+  buffer = toBuffer(buffer);
+  const zip = await JSZip.loadAsync(buffer);
+  const names: string[] = [];
+  const projections: { [key: string]: proj4.Converter } = {};
+  whiteList = whiteList || [];
+
+  for (const key of Object.keys(zip.files)) {
+    if (key.indexOf("__MACOSX") !== -1) {
+      continue;
+    }
+
+    const fileExt = key.slice(-4).toLowerCase();
+    const fileName = key.slice(0, -4);
+
+    if (fileExt === ".shp") {
+      names.push(fileName);
+      zip.files[fileName + ".shp"] = zip.files[key];
+    } else if (fileExt === ".prj") {
+      projections[fileName] = proj4(await zip.files[key].async("text"));
+    } else if (
+      fileExt === ".json" ||
```

Comment:
The length of `.json` is 5, so doesn't `fileExt` variable have the value?

Reply:
Done.